File: Task 1/C++/Components/Stock.cpp
#include "Stock.h"

Stock::Stock(){}

void Stock::addStock(Stock::stockItem_t* NewStock){
    if(NewStock->p == nullptr){return;};

    // Check For duplication of item
    if(getStock(NewStock->p->getName()) == nullptr){
        stockList.push_back(NewStock);
        return;
    }

    printf("ERROR: Item with Same name is already in stock\n");
}

void Stock::removeStock(const std::string stockName){
    // Search for item to remove
    for(int i = 0; i<stockList.size(); i++){
        if(stockList[i]->p->getName() == stockName){
            delete stockList[i]->p;
            delete stockList[i];
            stockList.erase(stockList.begin()+i);
            printf("\nDeleted Successfully\n");
            return;
        }
    }
    printf("\nERROR: Stock not found\n"); // not deleted
}

void Stock::updateStock(const std::string stockName, Stock::stockItem_t* NewStock){
    if(NewStock->p == nullptr){return;}
    Stock::stockItem_t* toUpdate = getStock(stockName); 
    if(toUpdate != nullptr){

         // Check that Transport either has the same code as the one being updated or a different code
        if ((NewStock->p->getName() == toUpdate->p->getName()) ||
            (NewStock->p->getName() != toUpdate->p->getName() && getStock(NewStock->p->getName()) == nullptr)) {
            removeStock(stockName);
            addStock(NewStock);
            printf("\nStock Updated\n");
            return;
        }
        printf("Stock was NOT Updated.\n");
    } else {
        printf("Stock not found.\n");
    }
}

void Stock::increaseStock(const std::string stockName, int newQuantity){
    
    // Search for item to increase stock
    Stock::stockItem_t* toIncrease = getStock(stockName); 
    if(toIncrease != nullptr){
        toIncrease->q += newQuantity;
        return;
    }
    printf("Stock not found.\n");
}

void Stock::clearStock(){
    // Memory Management
    for (Stock::stockItem_t* item : stockList) {
        delete item->p;
        delete item;
    }
    stockList.clear();
}

Stock::stockItem_t* Stock::getStock(const std::string stockName){
    for(stockItem_t* stock : stockList){
        if(stock->p->getName() == stockName){
            return stock;
        }
    }
    return nullptr;
}

void Stock::displayStock(){
    // Traverse and display
    std::cout << "Stock Items:\n";
    for (Stock::stockItem_t* item : stockList) {
        printf("\n%s", item->p->display_product_info().c_str());
        printf("\nQuantity: %s\n", std::to_string(item->q).c_str());
    }
}
void Stock::displayStock(const std::string stockName){
    // Traverse and display
    Stock::stockItem_t* item = getStock(stockName);
    if(item != nullptr){
        printf("\n%s", item->p->display_product_info().c_str());
        printf("\nQuantity: %s\n", std::to_string(item->q).c_str());
        return;
    }
    printf("\nERROR: Stock not found\n");
}

File: Task 2/Java/logistics/src/main/java/com/logistics/Components/Stock.java
package com.logistics.Components;

import java.util.ArrayList;
import java.util.List;

import com.logistics.Components.Product.Product;

public class Stock {
    public static class stockItem_t {
        public Product p;
        public int q;
        public stockItem_t(){this.p = null; this.q = 0;}
        public stockItem_t(Product p, int q){this.p = p; this.q = q;}
    }

    private List<stockItem_t> stockList;

    // Default constructor
    public Stock() {
        stockList = new ArrayList<>();
    }

    // Getter for the stock list
    public List<stockItem_t> getStockList() {return stockList;}

    // Adds a stock item to the stock list
    public void addStock(stockItem_t newStock) {
        if (newStock.p == null) {
            return;
        }

        // Check for duplication of item
        if (getStock(newStock.p.getName()) == null) {
            stockList.add(newStock);
            return;
        }

        System.out.println("ERROR: Item with the same name is already in stock");
    }

    // Removes a stock item from the stock list
    public void removeStock(String stockName) {
        // Search for the item to remove
        for (int i = 0; i < stockList.size(); i++) {
            if (stockList.get(i).p.getName().equals(stockName)) {
                stockList.remove(i);
                System.out.println("\nDeleted Successfully\n");
                return;
            }
        }
        System.out.println("\nERROR: Stock not found"); // not deleted
    }

    // Updates a stock item in the stock list
    public void updateStock(String stockName, stockItem_t newStock) {
        if (newStock.p == null) {
            return;
        }
        stockItem_t toUpdate = getStock(stockName);
        if (toUpdate != null) {
            // Check that the product either has the same name as the one being updated or a different name
            if (newStock.p.getName().equals(toUpdate.p.getName()) ||
               (!newStock.p.getName().equals(toUpdate.p.getName()) && getStock(newStock.p.getName()) == null)) {
                removeStock(stockName);
                addStock(newStock);
                System.out.println("\nStock Updated\n");
                return;
            }
            System.out.println("Stock was NOT Updated.\n");
        } else {
            System.out.println("Stock not found.\n");
        }
    }

    // Increases the quantity of a stock item in the stock list
    public void increaseStock(String stockName, int newQuantity) {
        // Search for the item to increase stock
        stockItem_t toIncrease = getStock(stockName);
        if (toIncrease != null) {
            toIncrease.q += newQuantity;
            return;
        }
        System.out.println("Stock not found.\n");
    }

    // Clears the stock list and manages memory
    public void clearStock() {
        // Memory Management
        for (stockItem_t item : stockList) {
            item.p = null;
        }
        stockList.clear();
    }

    // Retrieves a stock item from the stock list based on its name
    public stockItem_t getStock(String stockName) {
        for (stockItem_t stock : stockList) {
            if (stock.p.getName().equals(stockName)) {
                return stock;
            }
        }
        return null;
    }

    // Displays all stock items
    public void displayStock() {
        // Traverse and display
        System.out.println("Stock Items:");
        for (stockItem_t item : stockList) {
            System.out.printf("\n%s", item.p.display_product_info());
            System.out.printf("\nQuantity: %s\n", item.q);
        }
    }

    // Displays a specific stock item based on its name
    public void displayStock(String stockName) {
        // Traverse and display
        stockItem_t item = getStock(stockName);
        if (item != null) {
            System.out.printf("\n%s", item.p.display_product_info());
            System.out.printf("\nQuantity: %s\n", item.q);
            return;
        }
        System.out.println("\nERROR: Stock not found");
    }
}
     

